dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1231 @@
+      // (-6, -5, -4, -3, -2, -1) + 1 -> (-5, -4, -3, -2, -1, 0)
+      // (-5, -4, -3, -2, -1, 0) / 3 -> (-1, -1, -1, 0, 0, 0)
+      // Then shift back away from 0 (-1, -1, -1, 0, 0, 0) - 1 -> (-2, -2, -2, -1, -1, -1)
+      // Finally, with (num_channels - 1) we get the positive channel indices
+      // (-2, -1, 0, -2, -1, 0) + 3 - 1 -> (0, 1, 2, 0, 1, 2)
+      int reflect_dim_idx = (idx + 1) / num_channels - 1;
+      int inner_dim_idx = (idx + 1) % num_channels + num_channels - 1;
+      return border_remap(reflect_dim_idx, sample_shape.width, BorderTag<border>{}) * num_channels +
+             inner_dim_idx;
+    }
+    if (idx >= sample_shape.in_extents.x) {
+      return border_remap(idx / num_channels, sample_shape.width, BorderTag<border>{}) *
+                 num_channels +
+             idx % num_channels;
+    }
+    return idx;
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE In load(const In* __restrict__ in, ivec<axes> coords,
+                                         InShapeDesc<axes> sample_shape) const {
+    return in[dot(coords, sample_shape.in_strides)];
+  }
+
+ private:
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::REFLECT_101>) const {
+    assert(len > 0);
+    return boundary::idx_reflect_101(idx, len);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::REFLECT_1001>) const {
+    assert(len > 0);
+    return boundary::idx_reflect_1001(idx, len);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::CLAMP>) const {
+    assert(len > 0);
+    return boundary::idx_clamp(idx, 0, len);
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, int len,
+                                                  BorderTag<BoundaryType::WRAP>) const {
+    assert(len > 0);
+    return boundary::idx_wrap(idx, len);
+  }
+};
+
+template <typename In, int axes>
+class InLoaderPad {
+ public:
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap(int idx, InShapeDesc<axes> sample_shape,
+                                                  int axis) const {
+    (void)sample_shape;
+    (void)axis;
+    return idx;
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE int border_remap_innermost(int idx,
+                                                            InShapeDesc<axes> sample_shape) const {
+    (void)sample_shape;
+    return idx;
+  }
+
+  DALI_HOST_DEV DALI_FORCEINLINE In load(const In* __restrict__ in, ivec<axes> coords,
+                                         InShapeDesc<axes> sample_shape) const {
+    if (any_coord(coords < 0) || any_coord(coords >= sample_shape.in_extents)) {
+      return fill_value;
+    }
+    return in[dot(coords, sample_shape.in_strides)];
+  }
+
+  In fill_value;
```

Comment:
A scalar fill value is a bit underwhelming.

Reply:
I agree it may be useful. I just thought it is not a big priority as we already have operators that have scalar-only padding. As this PR grew considerbaly, I'd prefer making it a follow-up.

It should not be a big change on its own - if we enounter OOB access, we can retreive index channel (as we do for remapping in border reflect etc.) in the InLoader and take an input value accordingly.